src/Paprika/Merkle/RlpMemo.cs

```diff
@@ -64,139 +57,147 @@ public bool TryGetKeccak(byte nibble, out ReadOnlySpan<byte> keccak)
-        span.Clear();
-        var memo = new RlpMemo(span);
+        var nibbleIndex = index.SetCountBefore(nibble) - 1;
+        return _buffer.Slice(nibbleIndex * Keccak.Size, Keccak.Size);
+    }
 
-        // Extract empty bits if any
-        NibbleSet.Readonly empty;
+    private void GetIndex(out NibbleSet.Readonly index)
+    {
+        // Extract the index bits.
+        var indexLength = _buffer.Length % Keccak.Size;
 
-        // The empty bytes length is anything that is not aligned to the Keccak size
-        var emptyBytesLength = leftover.Length % Keccak.Size;
-        if (emptyBytesLength > 0)
+        if (indexLength != 0)
         {
-            var bits = leftover[^emptyBytesLength..];
-            NibbleSet.Readonly.ReadFrom(bits, out empty);
+            var bits = _buffer[^indexLength..];
+            NibbleSet.Readonly.ReadFrom(bits, out index);
         }
         else
         {
-            empty = default;
-        }
+            Debug.Assert(_buffer.Length is 0 or MaxSize);
 
-        var at = 0;
-        for (byte i = 0; i < NibbleSet.NibbleCount; i++)
-        {
-            if (children[i] && empty[i] == false)
-            {
-                var keccak = leftover.Slice(at * Keccak.Size, Keccak.Size);
-                at++;
-
-                memo.SetRaw(keccak, i);
-            }
+            index = _buffer.IsEmpty ? NibbleSet.Readonly.None : NibbleSet.Readonly.All;
         }
+    }
 
-        return memo;
-
-        [DoesNotReturn]
-        [MethodImpl(MethodImplOptions.NoInlining)]
-        static void ThrowDecompressionForbidden() => throw new InvalidOperationException("Decompression is forbidden.");
+    public static RlpMemo Copy(ReadOnlySpan<byte> from, scoped in Span<byte> to)
+    {
+        var span = to[..from.Length];
+        from.CopyTo(span);
+        return new RlpMemo(span);
     }
 
-    [SkipLocalsInit]
-    public static int Compress(in Key key, scoped in ReadOnlySpan<byte> memoizedRlp, NibbleSet.Readonly children, scoped in Span<byte> writeTo)
+    public static RlpMemo Insert(RlpMemo memo, byte nibble, ReadOnlySpan<byte> keccak, scoped in Span<byte> workingSet)
     {
-        // Optimization, omitting some of the branches to memoize.
-        // It omits only these with two children where the cost of the recompute is not big.
-        // To prevent an attack of spawning multiple levels of such branches, only even are skipped
-        if (children.SetCount == 2 && (key.Path.Length + key.StoragePath.Length) % 2 == 0)
-        {
-            return 0;
-        }
+        memo.GetIndex(out var index);
 
-        var memo = new RlpMemo(ComputeMerkleBehavior.MakeRlpWritable(memoizedRlp));
-        var at = 0;
+        // Ensure that this element doesn't already exist.
+        Debug.Assert(!index[nibble]);
 
-        var empty = new NibbleSet();
+        // Compute the destination size for copying.
+        var size = (index.SetCount < NibbleSet.NibbleCount - 1)
```

Comment:
Size computation is repeated into multiple places. Why don't we extract it as a method? It'd be based on the number of destination nibbles set right? So a simple static method would do?